src/jdk.internal.le/macosx/native/lible/CLibrary.cpp

```diff
@@ -0,0 +1,210 @@
+static jfieldID c_iflag;
+static jfieldID c_oflag;
+static jfieldID c_cflag;
+static jfieldID c_lflag;
+static jfieldID c_cc;
+static jfieldID c_ispeed;
+static jfieldID c_ospeed;
+
+static jclass winsize_j;
+static jfieldID ws_row;
+static jfieldID ws_col;
+static jfieldID ws_xpixel;
+static jfieldID ws_ypixel;
+
+static jclass nativelong_j;
+static jfieldID nativelong_value;
+
+JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_osx_CLibraryImpl_initIDs
+  (JNIEnv *env, jclass) {
+    jclass cls;
+    cls = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/LastErrorException");
+    CHECK_NULL(cls);
+    lastErrorExceptionClass = (jclass) env->NewGlobalRef(cls);
+    lastErrorExceptionConstructor = env->GetMethodID(lastErrorExceptionClass, "<init>", "(J)V");
+    CHECK_NULL(lastErrorExceptionConstructor);
+
+    cls = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/osx/CLibrary$termios");
+    CHECK_NULL(cls);
+    termios_j = (jclass) env->NewGlobalRef(cls);
+    CHECK_NULL(termios_j);
+    c_iflag = env->GetFieldID(termios_j, "c_iflag", "Ljdk/internal/org/jline/terminal/impl/jna/osx/NativeLong;");
+    CHECK_NULL(c_iflag);
+    c_oflag = env->GetFieldID(termios_j, "c_oflag", "Ljdk/internal/org/jline/terminal/impl/jna/osx/NativeLong;");
+    CHECK_NULL(c_oflag);
+    c_cflag = env->GetFieldID(termios_j, "c_cflag", "Ljdk/internal/org/jline/terminal/impl/jna/osx/NativeLong;");
+    CHECK_NULL(c_cflag);
+    c_lflag = env->GetFieldID(termios_j, "c_lflag", "Ljdk/internal/org/jline/terminal/impl/jna/osx/NativeLong;");
+    CHECK_NULL(c_lflag);
+    c_cc = env->GetFieldID(termios_j, "c_cc", "[B");
+    CHECK_NULL(c_cc);
+    c_ispeed = env->GetFieldID(termios_j, "c_ispeed", "Ljdk/internal/org/jline/terminal/impl/jna/osx/NativeLong;");
+    CHECK_NULL(c_ispeed);
+    c_ospeed = env->GetFieldID(termios_j, "c_ospeed", "Ljdk/internal/org/jline/terminal/impl/jna/osx/NativeLong;");
+    CHECK_NULL(c_ospeed);
+
+    cls = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/osx/CLibrary$winsize");
+    CHECK_NULL(cls);
+    winsize_j = (jclass) env->NewGlobalRef(cls);
+    ws_row = env->GetFieldID(winsize_j, "ws_row", "S");
+    CHECK_NULL(ws_row);
+    ws_col = env->GetFieldID(winsize_j, "ws_col", "S");
+    CHECK_NULL(ws_col);
+    ws_xpixel= env->GetFieldID(winsize_j, "ws_xpixel", "S");
+    CHECK_NULL(ws_xpixel);
+    ws_ypixel= env->GetFieldID(winsize_j, "ws_ypixel", "S");
+    CHECK_NULL(ws_ypixel);
+
+    nativelong_j = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/osx/NativeLong");
+    CHECK_NULL(nativelong_j);
+    nativelong_value = env->GetFieldID(nativelong_j, "value", "J");
+    CHECK_NULL(nativelong_value);
+}
+
+JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_osx_CLibraryImpl_tcgetattr
+  (JNIEnv *env, jobject, jint fd, jobject result) {
+    termios data;
+
+    if (tcgetattr(fd, &data) != 0) {
+        jobject exc = env->NewObject(lastErrorExceptionClass,
```

Comment:
Could lines 109-112 be converted to a utility function which would then be invoked at current lines 109, 145, 167, and 197?